### PR TITLE
More robust reading of filenames

### LINKED ISSUE #17: Robust filename parser
Issue extracted from a discussion in https://github.com/hannesm/patch/pull/9#discussion_r1541824996

Different implementations of diffs output different formats for the filename with spaces or special characters (such as backslashes, which are valid on Windows). For example with spaces:
- GNU diff:
  ```
  --- "a b"	2024-04-02 13:32:43.427214939 +0100
  +++ "a c"	2024-04-02 13:32:34.520202398 +0100
  @@ -1 +0,0 @@
  -test
  ```
- Git diff:
  ```
  diff --git a/a b b/a c
  index 039727e..e69de29 100644
  --- a/a b
  +++ b/a c
  @@ -1 +0,0 @@
  -test
  ```
- Busybox diff:
  ```
  --- a b
  +++ a c
  @@ -1 +0,0 @@
  -test
  ```

### PR DESCRIPTION
**Pull request by Leonidas-from-XIV**
This adds a lexer based on ocamllex to read the filenames of the patches.

Fixes #17.

### CODE REVIEW

**Comment by hannesm** [comment]
Thanks a lot, I'm curious about error conditions: 
- `--- a/foo "bar" baz`
- `--- "a/foo" bar baz`
- `--- /dev/null10`

Should these error out? Or is such a filename legal? I'd be happy to have them as part of the test suite (and I'm fine to consider them illegal and error out).

I'm asking since you clearly have more of the diff file format in your head than me. Thanks again.

**Comment by Leonidas-from-XIV** [comment]
Thanks for the edge cases. Coming up with these is sometimes tricky :-)

 * `--- a/foo "bar" baz` should be legal, I just ran the three tools and added their output to the tests. Looks like I have some patching to do.
 * `--- "a/foo" bar baz` I think that would imply that a file has a `/` in its name which is not possible on most filesystems so I'd say this is just strict wrong and should be rejected.
 * `--- /dev/null10` is a legal filename, but the parser will probably return `DevNull` by error here. I'll fix it.

**Comment by hannesm** [comment]
Thanks for your additional commits, what I had in mind is something with a starting quote:

```
diff --git a/test/test.ml b/test/test.ml
index 94f9e83..cb726a1 100644
--- a/test/test.ml
+++ b/test/test.ml
@@ -880,6 +880,17 @@ let busybox_diff_quotes = {|\
 let busybox_diff_quotes =
   operations [Patch.Edit ({|foo "bar" baz|}, {|foo bar "baz"|})] busybox_diff_quotes
 
+let busybox_diff_quotes2 = {|\
+--- "foo bar" baz
++++ "foo" bar "baz"
+@@ -1 +1 @@
+-This is wrong.
++This is right.
+|}
+
+let busybox_diff_quotes2 =
+  operations [Patch.Edit ({|"foo bar" baz|}, {|"foo" bar "baz"|})] busybox_diff_quotes2
+
 let filename_diffs =
   [
     "unified diff with spaces", `Quick, unified_diff_spaces;
@@ -888,6 +899,7 @@ let filename_diffs =
     "unified diff with quotes", `Quick, unified_diff_quotes;
     "git diff with quotes", `Quick, git_diff_quotes;
     "busybox diff with quotes", `Quick, busybox_diff_quotes;
+    "busybox diff with quotes 2", `Quick, busybox_diff_quotes2;
   ]
 
 let tests = [
```

And that fails at the moment with:
```
│ [FAIL]        filename                  3   unified diff with quotes.                                 │
└───────────────────────────────────────────────────────────────────────────────────────────────────────┘
ASSERT File "test/test.ml", line 811, characters 34-41
FAIL File "test/test.ml", line 811, characters 34-41

   Expected: `[--- a/foo "bar" baz
+++ b/foo bar "baz"
]'
   Received: `[--- a/foo \
+++ b/foo bar \
]'

```

**Comment by Leonidas-from-XIV** [comment]
True, there's an additional error case when the file starts with quotes. I've updated the tests so the destination filename starts with a quote.

I have a feeling that it might not be possible to read the filenames with spaces and quotes with just the lexer reading a string, given busybox and git diff do the opposite things when they encounter quotes and spaces (git diff quotes spaces and escapes quotes, busybox always uses literal quotes). Might need to properly tokenize and postprocess the outcome.

File: test/test.ml [hunk 1]
```diff
@@ -806,10 +806,106 @@ let unified_diff_creation = [
...
+let unified_diff_quotes = {|\
+--- "foo bar \"baz\""	2024-09-27 11:09:48.325541553 +0200
++++ "\"foo\" bar baz"	2024-09-27 11:06:42.612922437 +0200
+@@ -1 +1 @@
+-This is right.
++This is wrong.
+|}
+
+let unified_diff_quotes =
+  operations [Patch.Edit ({|foo "bar" baz|}, {|"foo" bar baz|})] unified_diff_quotes
```
Comment (hannesm) [inline test/test.ml:857]:
```suggestion
  operations [Patch.Edit ({|foo bar "baz"|}, {|"foo" bar baz|})] unified_diff_quotes
```

File: test/test.ml [hunk 2]
```diff
@@ -806,10 +806,106 @@ let unified_diff_creation = [
...
+index 88adca3..ef00db3 100644
+--- "a/foo bar \"baz\"" 
++++ "b/\"foo\" bar baz" 
+@@ -1 +1 @@
+-This is right.
++This is wrong.
+|}
+
+let git_diff_quotes =
+  operations [Patch.Edit ({|foo "bar" baz|}, {|"foo" bar baz|})] git_diff_quotes
```
Comment (hannesm) [inline test/test.ml:870]:
```suggestion
  operations [Patch.Edit ({|foo bar "baz"|}, {|"foo" bar baz|})] git_diff_quotes
```

File: test/test.ml [hunk 3]
```diff
@@ -806,10 +806,106 @@ let unified_diff_creation = [
...
+let busybox_diff_quotes = {|\
+--- foo bar "baz"
++++ "foo" bar baz
+@@ -1 +1 @@
+-This is right.
++This is wrong.
+|}
+
+let busybox_diff_quotes =
+  operations [Patch.Edit ({|foo "bar" baz|}, {|"foo" bar baz|})] busybox_diff_quotes
```
Comment (hannesm) [inline test/test.ml:881]:
```suggestion
  operations [Patch.Edit ({|foo bar "baz"|}, {|"foo" bar baz|})] busybox_diff_quotes
```

**Comment by hannesm** [comment]
Dear @Leonidas-from-XIV,

> True, there's an additional error case when the file starts with quotes. I've updated the tests so the destination filename starts with a quote.

Thanks. I added some comments on what I'd expect the outcome to be.

> I have a feeling that it might not be possible to read the filenames with spaces and quotes with just the lexer reading a string, given busybox and git diff do the opposite things when they encounter quotes and spaces (git diff quotes spaces and escapes quotes, busybox always uses literal quotes). Might need to properly tokenize and postprocess the outcome.

I have no file with a `"` on my system. I'd as well be fine to error out, and basically have the lexer check whether the entire string is consumed -- and if not, error.

If you strive for supporting both git diff and busybox, please also feel free to do so :)

My intuition is only to "better fail than use incomplete data" -- i.e. in the `"foo " bar` case. I'm happy to have a not 100% compatible implementation, and the first issue report where someone has a quote in their filename could serve as motivation to finish such an implementation.

File: src/patch.ml [hunk 4]
```diff
@@ -247,18 +247,29 @@ let pp ~git ppf {operation; hunks; mine_no_nl; their_no_nl} =
 let pp_list ~git ppf diffs =
   List.iter (Format.fprintf ppf "%a" (pp ~git)) diffs
 
+let process_git_prefix ~git ~prefix s =
+  if git && String.is_prefix ~prefix s then
+    String.slice ~start:(String.length prefix) s
+  else
+    s
+
```
Comment (kit-ty-kate) [inline src/patch.ml:255]: The prefix should never be taken out without input from the user. See https://github.com/hannesm/patch/pull/9
```suggestion
```

**Comment by Leonidas-from-XIV** [comment]
Generally I totally agree but I didn't want to add an unrelated change to the behavior to this PR which is just about the parsing of filenames.

File: src/patch.ml [hunk 5]
```diff
@@ -247,18 +247,29 @@ let pp ~git ppf {operation; hunks; mine_no_nl; their_no_nl} =
 let pp_list ~git ppf diffs =
   List.iter (Format.fprintf ppf "%a" (pp ~git)) diffs
 
+let process_git_prefix ~git ~prefix s =
```
Comment (kit-ty-kate) [inline src/patch.ml:250]: Fair enough. In that case i'd like to add a comment
```suggestion
(* TODO: remove this and let users decide the prefix level they want *)
let process_git_prefix ~git ~prefix s =
```

**Comment by Leonidas-from-XIV** [comment]
Making it pass `diff -u`, `git diff` and `busybox diff` output was a bit tricky.

There is still a corner case that I don't have an idea for: if the file is called `"foo bar"`, then `git diff` will generate `"\"foo bar\""` but busybox diff will generate `"foo bar"` and the unquoting logic does not know whether it needs to remove the outer quotes or not.

**Comment by kit-ty-kate** [comment]
Thanks a lot for this work! I have several ideas on how to improve this PR, i'll take over this work if that's ok with you, and i hope to be done on Monday or Tuesday night if all goes well

**Comment by kit-ty-kate** [comment]
I think this is ready to review. Don't mind the commit history, i'll squash and clean things before merging.

I strongly believe this supports everything GNU diff can throw at it and every other diff. Some OS have diffs that can output files that cannot be parsed by their own patch utility (e.g. if a filename contains a tab or a newline on any diff utility other than GNU or git it's pretty much game over) and for that we can't do anything. There is still the ambiguity between `"some quoted filename"` between GNU/git and the others but for that there isn't anything we can do either apart from having proper modes which i think would probably be a waste of time.

Once this PR is merged i'm hoping to:
- continue https://github.com/hannesm/patch/pull/9
- fix the printer when confronted with unprintable characters
- open a new PR to use the new `exception Error` in a lot more places because the current error modes are a bit all over the place (asserts, failure, invalid_args, …) and should probably be cleaned up before use in opam
- test the new code in https://github.com/ocaml/opam/pull/5892 and for every opam-repository packages with patches
- fix things if needed
- do a release

What do you think @hannesm?

**Comment by hannesm** [comment]
Dear Kate (and Marek),

I'm very happy of the work you're doing on this package. Incredibly happy. I briefly looked at the API change of this PR, and am mostly fine with it. The minor thing I find slightly annoying is the "exception Error" - I'd appreciate something that is more self-explanatory (i.e. 'exception Filename_parse_error of string' -- to include the "offending" data -- so if someone hits it, we easily get that information.

>     * continue [Add a ~p parameter to Patch.to_diffs mimicking the behaviour of patch -p<num> #9](https://github.com/hannesm/patch/pull/9)

Awesome!

>     * fix the printer when confronted with unprintable characters

Nice!

>     * open a new PR to use the new `exception Error` in a lot more places because the current error modes are a bit all over the place (asserts, failure, invalid_args, …) and should probably be cleaned up before use in opam

I agree that deciding on a single error mechanism is good, and `assert false` is not a good choice.

>     * test the new code in [Apply patches when dealing with repositories via an OCaml implementation of patch ocaml/opam#5892](https://github.com/ocaml/opam/pull/5892) and for every opam-repository packages with patches

\o/

>     * fix things if needed

yes, that'd be fantastic :D

>     * do a release

👯 👍🏾 

> What do you think @hannesm?

I'm happy. I've limited time for reviewing at the moment. If you're stuck / undecided on some decision, feel free to reach out to me.

Otherwise, feel free to push stuff forward - I think you have full permissions to this repository. This PR is great, adds tests and fixes issues that exist for real.

**Comment by kit-ty-kate** [comment]
I have a version cleanly rebased at https://github.com/kit-ty-kate/patch/commits/filename-spaces/
@Leonidas-from-XIV does that look good for you?

**Comment by hannesm** [comment]
> how does the last commit look?

looks good, thanks!

File: src/dune [hunk 6]
```diff
@@ -2,8 +2,9 @@
  (name patch)
  (synopsis "Patch purely in OCaml")
  (public_name patch)
- (modules patch)
- (wrapped false))
+ (modules patch patch_lib fname))
```
Comment (Leonidas-from-XIV) [inline src/dune:5]: My suggestion would be dropping this and just moving the executable into a different folder, thus new modules would be picked up automatically.

File: src/fname.mli [hunk 7]
```diff
@@ -0,0 +1,5 @@
+val parse : string -> (string option, string) result
+(** [parse s] parses [s] and returns a filename or [None] if the filename
+    is equivalent to [/dev/null].
+
+    Returns [Error msg] in case of error. *)
```
Comment (Leonidas-from-XIV) [inline src/fname.mli:5]: I am not sure if this is a clear API, when I get `Ok None` I am not sure I would think of `/dev/null`.

I know I started with this idea originally, but I think returning `Ok "/dev/null"` is better and then other tools on top can map "/dev/null" to whatever logic they find reasonable.

**Review by Leonidas-from-XIV — COMMENTED** [review]
Thanks for finishing and refactoring the PR. I have some small suggestions but it looks good overall and I think this gets as close to parsing a filename as we can get without getting extra info about the format to parse.

File: src/fname.mli (hunk 7 above)
Comment (kit-ty-kate) [reply src/fname.mli:5]: This module is not accessible outside of the patch library so i don't think we should rather make the internal API consistent with the rest of the code which expects an option (until someone proposes something better, although i'm personally fine with the current state of things)

File: src/dune (hunk 6 above)
Comment (kit-ty-kate) [reply src/dune:5]: this can be done later

**Comment by kit-ty-kate** [comment]
Thanks!
